### PR TITLE
Introduces a "Syntax Index" chapter to TRPL.

### PR DESCRIPTION
**Pull request by DanielKeep**
The intent with this chapter is to have a central place where users can
go to find out what a random bit of syntax means, be it a keyword,
symbol, or some unusual bit of composite syntax (like `for <...>`).  This
should be useful both for new users (who may not know what to call this
weird `'blah` thing), and for experienced users (who may just wish to
link someone to the appropriate section on `Trait + Trait` bounds).

Where possible, entries have been linked to an appropriate section of
the book which explains the syntax.  This was not possible in all cases.
If an entry is missing links, that's because I was unable to _find_
anything appropriate to link to.

This commit should include all stable keywords, operators and symbols,
as well as a selection of potentially confusing or unusual syntax.

### CODE REVIEW

**Comment by killercup** [comment]
This is great! When I was beginning to learn Rust I wished there was a list like this. (The reference is a bit long to easily search for stuff.)

File: src/doc/trpl/syntax-index.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+* `::path`: path relative to the crate root (*i.e.* an explicitly absolute path).  See [Crates and Modules (Re-exporting with `pub use`)].
+* `self::path`: path relative to the current module (*i.e.* an explicitly relative path).  See [Crates and Modules (Re-exporting with `pub use`)].
+* `super::path`: path relative to the parent of the current module.  See [Crates and Modules (Re-exporting with `pub use`)].
+* `type::ident`: associated constants, functions, and types.  See [Associated Types].
+* `<type>::…`: associated item for a type which cannot be directly named (*e.g.* `<&T>::…`, `<[T]>::…`, *etc.*).  See [Associated Types].
+
+<!-- Generics -->
+
+* `path<…>`: specifies parameters to generic type *in a type*.  See [Generics].
+* `path::<…>`: specifies parameters to generic type or function *in an expression*.
```
Comment (killercup) [inline src/doc/trpl/syntax-index.md]: Maybe expand the example a bit to show this can be used for functions, e.g. `"42".parse::<u32>()`.

**Comment by steveklabnik** [comment]
I like it. Thanks!

File: src/doc/trpl/syntax-index.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+* `for`: iterator loop, part of trait `impl` syntax, and higher-ranked lifetime syntax.  See [Loops (`for`)], [Method Syntax].
+* `if`: conditional branching.  See [`if`], [`if let`].
+* `impl`: inherent and trait implementation blocks.  See [Method Syntax].
+* `in`: part of `for` loop syntax.  See [Loops (`for`)].
+* `let`: variable binding.  See [Variable Bindings].
+* `loop`: unconditional, infinite loop.  See [Loops (`loop`)].
+* `match`: pattern matching.  See [Match].
+* `mod`: module declaration.  See [Crates and Modules (Defining Modules)].
+* `move`: part of closure syntax.  See [Closures (`move` closures)].
+* `mut`: denotes mutability in pointer types, pattern bindings, and `struct` fields.  See [Mutability].
```
Comment (petrochenkov) [inline src/doc/trpl/syntax-index.md]: s/, and `struct` fields//

File: src/doc/trpl/syntax-index.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+* `+=` (`var += expr`): arithmetic addition & assignment.
+* `,`: argument and element separator.  See [Attributes], [Functions], [Structs], [Generics], [Match], [Closures], [Crates and Modules (Importing Modules with `use`)].
+* `-` (`expr - expr`): arithmetic subtraction.  Overloadable (`Sub`).
+* `-` (`- expr`): arithmetic negation.  Overloadable (`Neg`).
+* `-=` (`var -= expr`): arithmetic subtraction & assignment.
+* `->` (`fn(…) -> type`, `|…| -> type`): function and closure return type.  See [Functions], [Closures].
+* `.` (`expr.ident`): member access.  See [Structs], [Method Syntax].
+* `..` (`..`, `expr..`, `..expr`, `expr..expr`): right-exclusive range literal.
+* `..` (`..expr`): struct literal update syntax.  See [Structs (Update syntax)].
+* `..` (`..ident`): "and the rest" pattern binding.  See [Patterns (Ignoring bindings)].
```
Comment (petrochenkov) [inline src/doc/trpl/syntax-index.md]: s/..ident/../
"the rest" pattern can't be bound to a name

File: src/doc/trpl/syntax-index.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+* `<` (`expr < expr`): less-than comparison.  Overloadable (`Cmp`, `PartialCmp`).
+* `<=` (`var <= expr`): less-than or equal-to comparison.  Overloadable (`Cmp`, `PartialCmp`).
+* `=` (`var = expr`, `ident = type`): assignment/equivalence.  See [Variable Bindings], [`type` Aliases], generic parameter defaults.
+* `==` (`var == expr`): comparison.  Overloadable (`Eq`, `PartialEq`).
+* `=>` (`pat => expr`): part of match arm syntax.  See [Match].
+* `>` (`expr > expr`): greater-than comparison.  Overloadable (`Cmp`, `PartialCmp`).
+* `>=` (`var >= expr`): greater-than or equal-to comparison.  Overloadable (`Cmp`, `PartialCmp`).
+* `>>` (`expr >> expr`): right-shift.  Overloadable (`Shr`).
+* `>>=` (`var >>= expr`): right-shift & assignment.
+* `@` (`expr @ expr`): pattern binding.  See [Patterns (Bindings)].
```
Comment (petrochenkov) [inline src/doc/trpl/syntax-index.md]: s/expr @ expr/ident @ expr/

**Comment by petrochenkov** [comment]
>When I was beginning to learn Rust I wished there was a list like this.

+1, this is very useful

File: src/doc/trpl/syntax-index.md (hunk 2 above)
Comment (DanielKeep) [inline src/doc/trpl/syntax-index.md]: \**stares*\*  Well, that's embarrassing.

**Comment by DanielKeep** [comment]
@steveklabnik I made some changes prompted by @petrochenkov's comments; hope that doesn't stuff anything up.  Good bors *pats*

**Comment by aochagavia** [comment]
Would it be worth it to add keywords that are reserved but not used (like `virtual`)?

**Comment by DanielKeep** [comment]
@aochagavia I decided against it on the basis that doing so can only confuse things.  Such keywords can't show up in existing code and can't be used in new code.  It'd be a bit like having keywords from unrelated languages with the note "not used in Rust at all in any way".

*That said*, it just occurred to me that you *could* make an argument for obsolete syntax.  We do have a *bit* of an issue with old code examples, so maybe it'd be worth listing old syntax at the end.  But not in this PR.  I don't want to create *even more* work for poor old Bors.  \**throws Bors a cookie*\*

**Comment by TyOverby** [comment]
> ! (expr!(…), expr!{…}, expr![…]): denotes macro expansion. See Macros.'

Shouldn't those be "ident" not "expr"?

**Comment by Gankro** [comment]
@DanielKeep I'd be interested in noting reserved (and deprecated) things, as code doesn't clearly indicate what version it works with. If you see `<-` and the docs you check don't mention it, you're just left going `????`. But if you see it's reserved, that's an indication you need newer docs.

**Comment by Diggsey** [comment]
`for` seems to link to "method syntax", which has no explanation of higher ranked lifetime syntax (or any other kind of "for")

**Comment by DanielKeep** [comment]
@Gankro My concern was that, for example, it's possible that `in` is going to change to `<-`, which would make old docs *just* as useless as not having anything for that.  I'd contend that if you see syntax that's not in the index, then *of course* you need newer docs.

Maybe there's a use for stuff in nightly that hasn't made it to stable yet, though.

**Comment by DanielKeep** [comment]
@Diggsey That's because I couldn't *find* any documentation on higher ranked lifetime syntax in the book.

**Comment by steveklabnik** [comment]
Yes, there's a bug open for adding HRTB

**Comment by supr** [comment]
@DanielKeep nice work :+1:. Love it, would end up helping a lot when searching for syntax information quickly.

**Comment by Gankro** [comment]
@Diggsey it just seems like all that needs to be done for that is to note when syntax is unstable, no?

**Comment by Diggsey** [comment]
@Gankro The `for<'a>` syntax isn't unstable though? (At least it works on stable rust!)

**Comment by Gankro** [comment]
@Diggsey sorry I meant to ping @DanielKeep

(dang tab completion!)

**Pull request merged**